### ISSUE TITLE
infinite [knob] (send delta values)

### ISSUE THREAD

**Issue by esaruoho** [porres/pd-else#1933]
there was an original claim that there's a limitless rotary encoder in knob and all i gotta do is click on the help and it's there.

there is no mention of Rotary or Endless in the documentation, and after bunches of back'n'forth in comments, it turns out there is no such thing available.

https://www.facebook.com/groups/puredata/posts/10160354837214495/

**Comment by porres** [comment]
so, is that really it? whenever you turn the knob, no matter how fast you turn it, like no matter what the delta is, you always get either "1" or "-1"? Or if you turn it really fast/wider you end up having a faster increment or decremet?

I had to stop and think where have I seen such a thing, and maybe the phone gain knob of my sound card is like that. Do you have any other examples? What exactly are you trying to implement in Pd with this?

**Comment by esaruoho** [comment]
Rotary encoders would tend to send +1 on every turn to the right and -1 on every turn to the left. So if u turn "Five steps" to the right you would be sending +1 Five times. Not +5.  Instead Five separate instances of +1. I hope that helps somewhat as info

**Comment by esaruoho** [comment]
I have had multiple devices midicontrollers with Rotary endless knobs and basically they send 63 and 64 out incessantly as many times as you turn it. This is for having incremental control over pitch parameters, like a radio tuning knob.

**Comment by esaruoho** [comment]
One device that i used to have is the Faderfox UC4 which has Rotary endless encoders and those r awesome

**Comment by porres** [comment]
and what do you need this in a patch for? you expect it to behave like a rotary knob by allowing the user to click and drag it around for that, right? Like, it would be the same or similar to using knob with its 'circular' mode, huh?

**Comment by esaruoho** [comment]
> and what do you need this in a patch for? you expect it to behave like a rotary knob by allowing the user to click and drag it around for that, right? Like, it would be the same or similar to using knob with its 'circular' mode, huh?

For Extremely slow pitch changes kind of like tuning a radio. I sold the uc4 and dont have a single Rotary encoder anymore in my setup so being able to have one or more within PD would be amazing. The problem with current circular is that it does not send +1 -1 it instead sends decimals all over the place, which i cannot use for anything.

**Comment by ben-wes** [comment]
@esaruoho : I use the rotary knobs of the Akai Mini Mk3 quite a lot and when I made patches for them, I realized that they not only send -1 and +1 (which are then encoded as 127 and 1), but also bigger values if you rotate faster. Decoding is easily done with a `moses 64` where the left output is the `inc` values and the right output with a `- 128` is the `dec` values.

With a little trick and an underlying `vsl`, you can also create the endless rotary behavior with the current `else/knob`. I'm pasting a little patch with this kind of abstraction below (didn't try with the Akai since I don't have it here right now).

@porres : Maybe this kind of vertical or horizontal interaction might be relevant. The relative values might also be an option. But on the other hand, this works quite simply - so not sure if additional features are necessary?

`#N canvas 426 206 665 597 12;
#X floatatom 62 259 6 0 0 0 - - - 0;
#X obj 62 118 moses 64;
#X obj 115 141 - 128;
#X obj 62 64 ctlin 0;
#N canvas 788 138 450 300 knob 0;
#X obj 100 102 vsl 31 26 -16 16 0 0 empty empty empty 0 -9 0 12 #fcfcfc #ffffff #000000 0 1;
#X obj 90 136 t f b;
#X obj 120 179 f;
#X obj 90 179 +;
#X msg 184 81 set 0;
#X obj 90 203 t f f;
#X obj 100 100 knob 31 0 127 0 0 \$0_out \$0_in #ffffff #ffffff black 0 1 0 0 0 360 0 0 0;
#X obj 184 222 mod 128;
#X obj 90 248 outlet;
#X obj 90 56 inlet;
#X connect 0 0 1 0;
#X connect 1 0 3 0;
#X connect 1 1 4 0;
#X connect 2 0 3 1;
#X connect 3 0 2 0;
#X connect 3 0 5 0;
#X connect 4 0 0 0;
#X connect 5 0 8 0;
#X connect 5 1 7 0;
#X connect 7 0 6 0;
#X connect 9 0 1 0;
#X coords 0 -1 1 1 31 31 2 100 100;
#X restore 62 193 pd knob;
#X connect 1 0 4 0;
#X connect 1 1 2 0;
#X connect 2 0 4 0;
#X connect 3 0 1 0;
#X connect 4 0 0 0;
`

**Comment by ben-wes** [comment]
... and this simpler version outputs relative and absolute values with the circular `else/knob` and its regular interaction:

```
#N canvas 520 149 450 466 12;
#X obj 70 60 knob 31 0 127 0 0 empty empty #ffffff #ffffff black 0 1 0 0 0 360 0 0 0;
#X obj 70 158 t f f;
#X obj 70 194 -;
#X obj 70 236 - 64;
#X obj 70 284 - 64;
#X obj 70 260 mod 128;
#X obj 70 134 i;
#X obj 199 327 +;
#X obj 229 327 f;
#X obj 70 308 t f f;
#X floatatom 199 377 5 0 0 0 - - - 0;
#X floatatom 70 377 5 0 0 0 - - - 0;
#X text 111 376 relative;
#X text 239 376 absolute;
#X connect 0 0 6 0;
#X connect 1 0 2 1;
#X connect 1 1 2 0;
#X connect 2 0 3 0;
#X connect 3 0 5 0;
#X connect 4 0 9 0;
#X connect 5 0 4 0;
#X connect 6 0 1 0;
#X connect 7 0 8 0;
#X connect 7 0 10 0;
#X connect 8 0 7 1;
#X connect 9 0 11 0;
#X connect 9 1 7 0;
```

(sorry for the pd syntax spam - don't know how to properly share patches here)

**Comment by porres** [comment]
> not sure if additional features are necessary?

I was wondering the same

**Comment by porres** [comment]
> (sorry for the pd syntax spam - don't know how to properly share patches here)

please upload it as a zipped file!

**Comment by ben-wes** [comment]
Ah, `.zip` makes sense - sorry, I didn't see that. Only tried `.pd` and `.txt`.

Both patches compressed here now:
* knob_experiments.zip](https://github.com/porres/pd-else/files/13260399/knob_experiments.zip)

**Comment by porres** [comment]
yup, there you go, kind of what I had in mind

**Comment by porres** [comment]
referenced in an open issue

**Comment by porres** [comment]
so, an infinite mode would only output relative change, huh? now that i fixed a long standing bug in circular dragging, I can finally try this

**Comment by porres** [comment]
> so, an infinite mode would only output relative change, huh? now that i fixed a long standing bug in circular dragging, I can finally try this

I don't think that is useful, I'm just offering the infinite increment right out of the knob's output, makes better sense, right? If anyone wants delta, you can just subtract from the last output...

Anyway, this is implemented as a "circular 2" mode... "circular 1" still exists as before and means "looping".

<img width="284" alt="Image" src="https://github.com/user-attachments/assets/18facab7-654b-49d9-b350-5ef6f5c552ae" />

**Comment by esaruoho** [comment]
> So, [@ben-wes](https://github.com/ben-wes) and [@esaruoho](https://github.com/esaruoho) you can test the new circular mode in https://github.com/porres/pd-else/actions/runs/15892189720
> 
> And Esa, sorry I took this long. I was stuck with these circular dragging bugs forever since the beggining and I was finally able to solve them! Hope this is still useful to you

if i can get it to work, it'll definitely be very very useful! thanks for looking into it!

**Comment by porres** [comment]
And I haven't forgotten about bicoeff, it's just that I'm still very very bad with tcl/tk :)
